functions/main.py

```diff
@@ -168,3 +168,20 @@ def add_custom_claim_for_the_plan(event: ScheduledEvent) -> None:
             auth.set_custom_user_claims(plan.id, {"plan": "basic"})
         else:
             auth.set_custom_user_claims(plan.id, {"plan": "pro"})
+
+
+@on_document_written(document="plans/{planId}")
+def update_user_plan(event: Event[DocumentSnapshot | None]) -> None:
+    logging.info("Running update_user_plan")
+    user_id = event.params["planId"]
+    try:
+        auth.get_user(user_id)
+    except auth.UserNotFoundError:
+        return None
+    if event.data is None or event.data.after is None or (data := event.data.after.to_dict()) is None:
+        auth.set_custom_user_claims(user_id, {"plan": "basic"})
+    elif not("endDate" in data) or int(data.get("endDate").timestamp()) < get_int_utcnow():
```

Comment:
let's invert this logic, since there's a bug, if there's no `endDate` in `data` we set the plan to pro for ever.

So let's do:

```python
if "endDate" in data and int(data.get("endDate").timestamp()) > get_int_utcnow():
    auth.set_custom_user_claims(user_id, {"plan": "pro"})
    return
```

then at the end, set to the plan basic.